subprojects/model-core/src/main/java/org/gradle/internal/instantiation/generator/AbstractClassGenerator.java

```diff
@@ -620,7 +620,11 @@ public String getName() {
         }
 
         public boolean isReadOnly() {
-            return mainGetter != null && setters.isEmpty();
+            return isReadable() && !isWritable();
+        }
+
+        public boolean isReadableAndNotDirectlyWritable() {
+            return isReadable() && setters.stream().noneMatch(method -> method.getParameterTypes()[0].equals(getType()));
```

Comment:
It think that using `equals` should be fine here: We are only looking at `Property` types here anyway, so the raw setter type should be the same as the container type.